### PR TITLE
ARTEMIS-1607 OpenWire is sending responses too early with durable messages

### PR DESCRIPTION
**Pull request by franz1981**
AMQSession is sending response back without waiting past I/O tasks on StorageManager to be finished

### CODE REVIEW

**Comment by franz1981** [comment]
Possibly I will add a test but is tricky and I need to run CI on it too :+1:

**Comment by clebertsuconic** [comment]
The errors on the completion are critical.  They are IOerrors.  I think the critical exception will be called anyways.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQSession.java [hunk 1]
```diff
@@ -393,26 +395,47 @@ public void send(final ProducerInfo producerInfo,
...
+                           final Throwable criticalError = new IOException(errorMessage);
+                           try {
+                              //it is handled async and hopefully will be sent before the critical error shutdown the broker:
+                              //it helps to fail fast the clients on critical errors
+                              connection.serviceException(criticalError);
+                           } catch (Exception e) {
+                              ActiveMQServerLogger.LOGGER.debug(e);
+                           } finally {
+                              //it needs to be called ASAP: the broker isn't in a safe state
+                              server.getStorageManager().criticalError(criticalError);
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQSession.java]: this is probably also called earlier by the IO layer.. but it doesn't hurt here.

**Comment by clebertsuconic** [comment]
@franz1981 I told you it was ok to do it this way.. but I'm a bit concerned by the use of setResponse(true) and false through this.

is there a way to return the packet like it used to be done before? can you check the setRespond(true) and false usages? I'm concerned of a possible race making the server unavailable eventually.

**Comment by clebertsuconic** [comment]
@franz1981 typo on my last message.. I actually meant the autoRead stuff.

**Comment by franz1981** [comment]
@clebertsuconic Np I've understood what you mean :) 
I've pushed another commit with another solution less "reactive" but more simlar to the original code: let me know if it seems better and I will squash the commits into this last one :+1:

**Comment by franz1981** [comment]
@clebertsuconic I've reverted with the original not blocking version, but avoiding calling the shutdown, makes sense?
I'm just worried about racing accesses to the connection by different threads...

**Pull request closed without merging**